### ISSUE TITLE
plot

### ISSUE THREAD

**Issue by wenwendou** [chenning0115/SVAFormer#1]
Hello, I'm currently using your method as a comparison method. Your code only predicts the truth value part when generating the graph. Now I want to predict the entire graph, load the model and the PCA used during model training, and keep other parameters consistent with the model. Why is there only three or five color classifications when generating the graph? Here is the code I added below

    image, gt , _, _ = dataloader.load_data()  # raw_data: [C, H, W]

    # 2. 加载训练时保存的 PCA 模型
    pca = joblib.load(f'{dataset_n}pca_model.pkl')

    # 3. 对 image 应用 PCA 降维
    H, W, C = image.shape
    image_reshaped = image.reshape(-1, C)
    image_pca = pca.transform(image_reshaped).reshape(H, W, -1)  # [H, W, numComponents]

    # 4. 转置成模型输入格式 [C, H, W]
    image_pca = image_pca.transpose(2, 0, 1).astype(np.float32)

    # 5. 转标签格式为 int32
    gt = gt.astype(np.int32)
    best_model_path = os.path.join(trainer.checkpoint_dir, "model_best.pth")
    trainer.net.load_state_dict(torch.load(best_model_path, map_location=trainer.device))
    trainer.net.eval()
    # 6. 调用滑窗预测+绘图函数
    get_color_map.get_cls_map(
        net=trainer.net.to(trainer.device),
        device=trainer.device,
        image_data=image_pca,
        y_gt=gt,
        patch_size=19,  # 根据你模型的patch大小调整
        dataset_name=dataset_n  # 比如 "sl"
    )

# 预测标签映射回图像
def get_classification_map(y_pred, y):
    height, width = y.shape
    k = 0
    cls_labels = np.zeros((height, width))
    for i in range(height):
        for j in range(width):
            if y[i, j] == 0:
                continue
            cls_labels[i][j] = y_pred[k] + 1
            k += 1
    return cls_labels

# 不同数据集的配色表
def get_color_map(dataset_name):
    dataset_name = dataset_name.lower()

    salinas_colors = [
        [0, 0, 0], [176, 47, 95], [10, 254, 255], [255, 0, 255], [159, 32, 240],
        [127, 255, 213], [0, 139, 138], [4, 205, 0], [101, 174, 255], [255, 255, 0],
        [255, 0, 0], [215, 191, 216], [254, 127, 80], [160, 81, 45],
        [192, 192, 192], [218, 112, 214], [0, 96, 255]
    ]

    houston_colors = [
        [0, 0, 0], [255, 255, 0], [255, 0, 0], [0, 255, 0], [0, 0, 255],
        [255, 0, 255], [0, 255, 255], [128, 0, 0], [0, 128, 0], [0, 0, 128],
        [128, 128, 0], [128, 0, 128], [0, 128, 128], [192, 192, 192],
        [128, 128, 128], [64, 64, 64], [192, 0, 0], [0, 192, 0], [0, 0, 192],
        [192, 192, 0], [192, 0, 192], [0, 192, 192], [64, 0, 128]
    ]

    houston18_colors = [
        [0, 0, 0], [176, 47, 95], [10, 254, 255], [255, 0, 255], [159, 32, 240],
        [127, 255, 213], [0, 139, 138], [4, 205, 0], [101, 174, 255], [255, 255, 0],
        [255, 0, 0], [215, 191, 216], [254, 127, 80], [160, 81, 45],
        [192, 192, 192], [218, 112, 214], [0, 96, 255], [0, 255, 127],
        [255, 165, 0], [60, 179, 113], [123, 104, 238]
    ]

    if dataset_name == "sl":
        return salinas_colors
    elif dataset_name == "hh":
        return houston_colors
    elif dataset_name == "h18":
        return houston18_colors
    else:
        raise ValueError(f"Unknown dataset name: {dataset_name}")

# 标签映射到 RGB 颜色
def list_to_colormap(x_list, dataset_name):
    color_table = get_color_map(dataset_name)
    y = np.zeros((x_list.shape[0], 3))
    for index, item in enumerate(x_list):
        if item < len(color_table):
            y[index] = np.array(color_table[int(item)]) / 255.
        else:
            y[index] = np.array([0, 0, 0])
    return y

# 显示并保存分类图
def classification_map(map, ground_truth, dpi, save_path):
    fig = plt.figure(frameon=False)
    fig.set_size_inches(ground_truth.shape[1]*2.0/dpi, ground_truth.shape[0]*2.0/dpi)

    ax = plt.Axes(fig, [0., 0., 1., 1.])
    ax.set_axis_off()
    ax.xaxis.set_visible(False)
    ax.yaxis.set_visible(False)
    fig.add_axes(ax)

    ax.imshow(map)
    fig.savefig(save_path, dpi=dpi)
    plt.close(fig)
    return 0

def predict_whole_image(net, image_data, patch_size, device):
    pad_size = patch_size // 2
    C, H, W = image_data.shape

    # 对空间维度padding，使用reflect保证边界平滑
    padded = np.pad(image_data, ((0, 0), (pad_size, pad_size), (pad_size, pad_size)), mode='reflect')

    # 预测图初始化，注意边界中心点不可预测，先全置0（或背景类别）
    prediction_map = np.zeros((H, W), dtype=np.uint8)

    with torch.no_grad():
        for i in tqdm(range(H)):
            for j in range(W):
                # 取patch，注意padded索引从i开始，所以中心点对应prediction_map索引是 i,j
                patch = padded[:, i:i+patch_size, j:j+patch_size]  # [C, P, P]
                patch = torch.from_numpy(patch).unsqueeze(0).to(device).float()  # [1, C, P, P]

                output = net(patch)  # output: [1, num_classes]
                if isinstance(output, tuple):
                    output = output[0]

                pred = output.argmax(dim=1).item()

                # 赋值给中心点坐标
                center_i = i + pad_size
                center_j = j + pad_size

                # 注意这里prediction_map大小是H x W，对应image_data无padding索引
                # 因为patch的左上角对应padded的i,j，中心点对应padded的i+pad_size,j+pad_size
                # 但prediction_map无padding，所以索引需要减去pad_size
                if pad_size <= center_i < H + pad_size and pad_size <= center_j < W + pad_size:
                    prediction_map[center_i - pad_size, center_j - pad_size] = pred + 1

    return prediction_map

# def predict_whole_image(net, image_data, patch_size, device):
#
#     pad_size = patch_size // 2
#     C, H, W = image_data.shape
#
#     # 对空间维度padding
#     padded = np.pad(image_data, ((0, 0), (pad_size, pad_size), (pad_size, pad_size)), mode='reflect')
#     prediction_map = np.zeros((H, W), dtype=np.uint8)
#
#     with torch.no_grad():
#         for i in tqdm(range(H), desc='Predicting full image'):
#             for j in range(W):
#                 patch = padded[:, i:i+patch_size, j:j+patch_size]  # [C, P, P]
#                 patch = torch.from_numpy(patch).unsqueeze(0).to(device).float()  # [1, C, P, P]
#
#                 output = net(patch)  # output形状: [1, num_classes]
#                 # print(output.shape)
#                 pred = output.argmax(dim=1).item()
#
#                 prediction_map[i, j] = pred + 1
#
#     return prediction_map

# 主函数：预测整图并保存图像
def get_cls_map(net, device, image_data, y_gt, patch_size, dataset_name):
    net.eval()

    cls_labels = predict_whole_image(net, image_data, patch_size, device)
    print("Predicted unique labels:", np.unique(cls_labels))
    print("Ground truth unique labels:", np.unique(y_gt))
    x = np.ravel(cls_labels)
    gt = y_gt.flatten()

    y_list = list_to_colormap(x, dataset_name)
    y_gt_list = list_to_colormap(gt, dataset_name)

    y_pred_rgb = y_list.reshape((y_gt.shape[0], y_gt.shape[1], 3))
    y_gt_rgb = y_gt_list.reshape((y_gt.shape[0], y_gt.shape[1], 3))

    classification_map(y_pred_rgb, y_gt, dpi=300, save_path=f'classification_maps/{dataset_name}_full_predictions.png')
    classification_map(y_gt_rgb, y_gt, dpi=300, save_path=f'classification_maps/{dataset_name}_gt.png')

    print('------Full image classification maps saved-------')
